content/grants/persistent-memory.md

```diff
@@ -0,0 +1,158 @@
+
+## Concurrency
+
+Because nouns are immutable, and because the copy-on-write approach ensures that a new, dirty page is a copy of an existing clean page prior to mapping it, readers should never see an inconsistent state of an object so long as that object is traceable from a root.
+
+It is not necessary to support multiple writers to the arena. Event (`+poke`) computation is linearized: we cannot start computing the next event until the result of the previous event is copied to the arena. Computation of scries (`+peek`) may run concurrently, but will only read from the arena unless new code generation is necessary.
+
+Thus, it should suffice to maintain a single, global writer lock for an arena.
+Contention for this lock is not likely to be high, and this will eliminate a great deal of complexity which would obtain if page dirtying and commits required more granular synchronization.
+
+## Tuning
+
+Commit frequency of the Arvo snapshot arena is a potentially tunable value, with performance implications.
+Committing more often incurs more disk IO, saving nouns that will potentially soon be discarded (Ames queues, for instance). However, under memory pressure, it makes paging more performant, as the proportion of dirty pages which must be *written* to disk to evict is smaller. Already-committed pages can simply be evicted from memory and re-read from disk when a read fault occurs. Committing less often requires writes as part of eviction, but reduces disk IO. It is desirable as long as memory pressure as low, and highly desirable for block storage where write-wearing is a concern.
+
+
+# Evaluation
+An important part of this work is to objectively evaluate the proposed approach. This will require the implementation of at least one alternate approach. The primary alternative is to map pages `MAP_PRIVATE` and simply track dirty pages and set `PROT_WRITE` in the fault handler. Then dirty pages could be more lazily written to disk, using `write()`, and committing using `fsync()`. The main reason this tradeoff is not expected to perform well is that it requires the underlying OS to have swap space configured and available if a dirty page must be evicted from the page cache. Another way of stating this is that the size of our dirty page set is limited by available RAM and swap space.
+
+However, this approach is used both by the current Vere as well as by LMDB for writing dirtied pages. It should therefore be implemented for New Mars at least by way of comparison.
+
+# Milestones & Compensation
+
+## Milestone 1: Basic copy-on-write arena
+
+Implement a copy-on-write arena by using `write()` to copy pages to new regions of the backing file.
+This should include at least single-arena snapshot support.
+
+Compensation: 1 star
+
+## Milestone 2: Kernel copy-on-write (`MAP_PRIVATE`) + lazy write arena
+
+Implement a copy-on-write arena by mapping pages `MAP_PRIVATE` and lazily writing dirty pages to the backing file.
```

Comment:
It might be worth noting that this is will shortly be vere's approach.

Reply:
There is a difference in that the specified approach for committing here is to find free blocks and `write()` to them, then update the index. Vere `write()`s to a patch file which is later unified with the main snapshot memory image, correct?

Reply:
Interesting, I hadn't thought about the space between the main proposal and vere's approach. Don't mind me ...